### PR TITLE
User details

### LINKED ISSUE #21: User detail page (and update page)
### Feature request
------------------------
- Links to from account button in dropdown on global toolbar
- `/users/:id/update`
- If attempting to view someone else's id, 404 (Just display the Notfound page (or redirect to the not found page))
- Shows user info
- Allows user to edit their account information and save it (displayname, etc)

### PR DESCRIPTION
**Pull request by Searay-330**
## Proposed changes
--------------------
form used to show a user info and update it

## Types of changes
--------------------
What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist
--------------------
_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. This is simply a reminder of what we are going to look for before merging your code._

- [X] My changeset covers only what is described above (no extraneous changes)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged downstream

## Further comments
--------------------
i have not done error handling or user feedback i will be making an alert component for use but since this is dependent on it reviewing the main features is still ready

### CODE REVIEW

**Comment by Searay-330** [comment]
this PR also now comes with a universal alert/feedback system

**Comment by MikeDrewitt** [comment]
run `npm run format` please then push

File: src/components/pages/userDetailPage.tsx [hunk 1]
```diff
@@ -1,7 +1,44 @@
...
+const UserDetailPage = ({}) => {
+  const { userId } = useParams() as Params
+
+  const [loading, setLoading] = useState(true)
+  const [user, setUser] = useState({} as User)
+
+  useEffect(() => {
+    RequestService.get(`/api/users/${userId}`)
+      .then(setUser)
+      .catch((e) => console.error(e))
```
Comment (MikeDrewitt) [inline src/components/pages/userDetailPage.tsx]: this should catch the error, then set it. Then return `<ErrorPage error={error} />` if error is not null. Take a look at the `UserCoursesListPage` to see what that looks like.

File: src/components/pages/userDetailPage.tsx [hunk 2]
```diff
@@ -1,7 +1,44 @@
...
+      .finally(() => setLoading(false))
+  }, [])
+
+  if (loading) return <LoadingOverlay />
 
-const UserDetailPage = ({}) => <PageWrapper>User Detail Update</PageWrapper>
+  return (
+    <PageWrapper>
+      <div className={styles.userDetailDiv}>
```
Comment (MikeDrewitt) [inline src/components/pages/userDetailPage.tsx]: Idk how to feel about this class name.

File: src/components/shared/forms/editUserForm.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+import RequestService from 'services/request.service'
+
+type Props = {
+  user: User
+}
+
+const EditUserForm = ({ user }: Props) => {
+  const [formData, setFormData] = useState<User>(user)
+  const setAlert = useAppDispatch<typeof SET_ALERT>(SET_ALERT)
```
Comment (MikeDrewitt) [inline src/components/shared/forms/editUserForm.tsx]: You can just change this to 

```javascript
const setAlert = useAppDispatch(SET_ALERT)
```

File: src/components/shared/forms/editUserForm.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+  const handleUpdatePreferredName = (preferredName: string) => setFormData({ ...formData, preferredName })
+
+  const submit = () => {
```
Comment (MikeDrewitt) [inline src/components/shared/forms/editUserForm.tsx]: to follow the react convention, this should probably be called `handleSubmit` rather than just submit

File: src/components/shared/forms/editUserForm.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    RequestService.put(`/api/users/${user.id}`, formData)
+      .then((response) => {
+        const message = response.message
+        setAlert({ autoDelete: true, type: 'success', message })
```
Comment (MikeDrewitt) [inline src/components/shared/forms/editUserForm.tsx]: I think setting an explicit message here might be a bit nicer. Maybe something like `User has been updated`, idk.

Comment (MikeDrewitt) [reply src/components/shared/forms/editUserForm.tsx]: Also this as a one liner might be a bit cleaner since you don't really _have_ to declare message. Declaring it is a bit slower since it needs to allocate the memory for it just to let it get garbage collected.

File: src/components/shared/alerts/alert.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  const colors = getCssVariables()
+
+  let backgroundColor = colors.red
+
+  if (alert.type === 'warning') backgroundColor = colors.yellow
+  else if (alert.type === 'info') backgroundColor = colors.purple
+  else if (alert.type === 'success') backgroundColor = colors.green
+
+  return (
+    <>
```
Comment (MikeDrewitt) [inline src/components/shared/alerts/alert.tsx]: You don't need this fragment.

File: src/components/shared/alerts/alert.tsx [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import { useAppDispatch, useAppSelector } from 'redux/hooks'
+import { SET_ALERT } from 'redux/types/active.types'
+
+import { getCssVariables } from 'utils/theme.utils'
+
+import styles from './alert.scss'
+
+const Alert = () => {
+  const alert = useAppSelector((state) => state.active.alert)
+  const setAlert = useAppDispatch<typeof SET_ALERT>(SET_ALERT)
```
Comment (MikeDrewitt) [inline src/components/shared/alerts/alert.tsx]:
```javascript
const setAlert = useAppDispatch(SET_ALERT)
```

File: src/components/shared/alerts/alert.scss [hunk 8]
```diff
@@ -0,0 +1,102 @@
+.notification_container {
+  font-size: 14px;
+  box-sizing: border-box;
+  position: fixed;
+  z-index: 1;
+
+  .notification {
```
Comment (MikeDrewitt) [inline src/components/shared/alerts/alert.scss]: I thought you were gonna put this guy at the bottom left, rather than the top? Did you change your mind?

**Review by MikeDrewitt — CHANGES_REQUESTED** [review]

File: src/components/shared/alerts/alert.scss (hunk 8 above)
Comment (Searay-330) [reply src/components/shared/alerts/alert.scss]: you are correct i just forgot to carry the styles into the sheet from the browser

File: src/components/shared/forms/editUserForm.tsx (hunk 4 above)
Comment (Searay-330) [reply src/components/shared/forms/editUserForm.tsx]: changed

File: src/components/shared/forms/editUserForm.tsx (hunk 3 above)
Comment (Searay-330) [reply src/components/shared/forms/editUserForm.tsx]: changed

File: src/components/pages/userDetailPage.tsx (hunk 2 above)
Comment (Searay-330) [reply src/components/pages/userDetailPage.tsx]: updated it to something i think is more descriptive but its still just a how you feel about class names sort of thing

File: src/components/pages/userDetailPage.tsx (hunk 1 above)
Comment (Searay-330) [reply src/components/pages/userDetailPage.tsx]: yes this does mean that this is now setup where the error page display is not recoverable

File: src/components/shared/alerts/alert.tsx (hunk 6 above)
Comment (Searay-330) [reply src/components/shared/alerts/alert.tsx]: deleted it

File: src/components/shared/forms/editUserForm.tsx (hunk 5 above)
Comment (Searay-330) [reply src/components/shared/forms/editUserForm.tsx]: set a new hard coded message

File: src/components/pages/userDetailPage.tsx [hunk 9]
```diff
@@ -1,7 +1,48 @@
...
+  const { userId } = useParams() as Params
+
+  const [loading, setLoading] = useState(true)
+  const [user, setUser] = useState({} as User)
+  const [error, setError] = useState(null)
+
+  useEffect(() => {
+    RequestService.get(`/api/users/${userId}`)
+      .then(setUser)
+      .catch((e) => setError(e))
```
Comment (MikeDrewitt) [inline src/components/pages/userDetailPage.tsx]:
```suggestion
.catch(setError)
```

File: src/components/shared/forms/editUserForm.tsx [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+const EditUserForm = ({ user }: Props) => {
+  const [formData, setFormData] = useState<User>(user)
+  const setAlert = useAppDispatch(SET_ALERT)
+
+  const handleUpdatePreferredName = (preferredName: string) => setFormData({ ...formData, preferredName })
+
+  const handleSubmit = () => {
+    RequestService.put(`/api/users/${user.id}`, formData)
+      .then(() => {
+        setAlert({ autoDelete: false, type: 'success', message: 'User Preferences Updated' })
```
Comment (MikeDrewitt) [inline src/components/shared/forms/editUserForm.tsx]: You don't need to wrap this in brackets, imo it's a bit cleaner not being wrapped.

File: src/components/shared/forms/editUserForm.tsx [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+      <h1>User Information</h1>
+      <TextField
+        onChange={handleUpdatePreferredName}
+        label='Preferred Name'
+        id='preferredName'
+        placeholder={user.preferredName}
+      />
+      <TextField type='email' label='Email' id='email' placeholder={user.email} disabled />
+      <TextField label='Person Number' id='externalId' placeholder={user.externalId} disabled />
+      <button onClick={handleSubmit}>Submit</button>
```
Comment (MikeDrewitt) [inline src/components/shared/forms/editUserForm.tsx]: We should probably make this a button component. That way we're not using raw buttons anywhere

File: src/components/shared/alerts/alert.tsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+  let backgroundColor = colors.red
+
+  if (alert.type === 'warning') backgroundColor = colors.yellow
+  else if (alert.type === 'info') backgroundColor = colors.purple
+  else if (alert.type === 'success') backgroundColor = colors.green
+
+  return (
+    <div className={styles.notification_container}>
+      <div className={`${styles.notification} ${styles.alert}`} style={{ backgroundColor }}>
```
Comment (MikeDrewitt) [inline src/components/shared/alerts/alert.tsx]: Does this need to be 2 css classes? if they're both only being used once, why not just make them one class?

File: src/components/shared/alerts/alert.tsx [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+  const colors = getCssVariables()
+
+  let backgroundColor = colors.red
+
+  if (alert.type === 'warning') backgroundColor = colors.yellow
+  else if (alert.type === 'info') backgroundColor = colors.purple
+  else if (alert.type === 'success') backgroundColor = colors.green
+
+  return (
+    <div className={styles.notification_container}>
```
Comment (MikeDrewitt) [inline src/components/shared/alerts/alert.tsx]: Camel case class names, no snake case.

File: src/components/shared/alerts/alert.tsx [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+      <div className={`${styles.notification} ${styles.alert}`} style={{ backgroundColor }}>
+        <button onClick={handleRemoveAlert}>X</button>
+        <div>
+          <p className={styles.notification_message}>{alert.message}</p>
```
Comment (MikeDrewitt) [inline src/components/shared/alerts/alert.tsx]: camel case

File: src/components/shared/alerts/alert.tsx [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  let backgroundColor = colors.red
+
+  if (alert.type === 'warning') backgroundColor = colors.yellow
+  else if (alert.type === 'info') backgroundColor = colors.purple
+  else if (alert.type === 'success') backgroundColor = colors.green
+
+  return (
+    <div className={styles.notification_container}>
+      <div className={`${styles.notification} ${styles.alert}`} style={{ backgroundColor }}>
+        <button onClick={handleRemoveAlert}>X</button>
```
Comment (MikeDrewitt) [inline src/components/shared/alerts/alert.tsx]: This should use an `<FaIconButton />` instead of a button with an X in it. Then we can probably ditch all the styles around buttons in this scss file

**Review by MikeDrewitt — CHANGES_REQUESTED** [review]
Found some more stuff

File: src/components/pages/userDetailPage.scss [hunk 16]
```diff
@@ -0,0 +1,19 @@
+@import 'variables';
+
+.userInformationFormWrapper{
+	display: flex;
+	justify-content: center;
+}
+
+input::placeholder{
```
Comment (MikeDrewitt) [inline src/components/pages/userDetailPage.scss]: We probably should style the input here either. If we want to style it, we should probably do so at the component level so that all inputs are the same across pages.

File: src/components/pages/userDetailPage.tsx [hunk 17]
```diff
@@ -1,7 +1,48 @@
...
+      .finally(() => setLoading(false))
+  }, [])
+
```
Comment (MikeDrewitt) [inline src/components/pages/userDetailPage.tsx:36]:
```suggestion
if (loading) return <LoadingOverylay delay={250} />
```

File: src/components/pages/userDetailPage.tsx [hunk 18]
```diff
@@ -1,7 +1,48 @@
...
+  if (loading) return <LoadingOverlay />
```
Comment (MikeDrewitt) [inline src/components/pages/userDetailPage.tsx]:
```suggestion
if (loading) return <LoadingOverlay delay={250} />
```

File: src/components/shared/forms/editUserForm.tsx (hunk 11 above)
Comment (MikeDrewitt) [reply src/components/shared/forms/editUserForm.tsx]: Oh one more thing we might want to add is a semi transparent loading overlay for when this request is saving. That way they can only click save once and there's clearly something happening for them on save

File: src/components/shared/forms/editUserForm.tsx (hunk 10 above)
Comment (MikeDrewitt) [reply src/components/shared/forms/editUserForm.tsx]: Also I think this would probably be better as an autoDeleted message

File: src/components/shared/layouts/pageWrapper.tsx [hunk 19]
```diff
@@ -12,6 +13,7 @@ type Props = {
 const PageWrapper = ({ children, className = '' }: Props) => (
   <div className={styles.page}>
     <GlobalToolbar />
+    <Alert />
```
Comment (MikeDrewitt) [inline src/components/shared/layouts/pageWrapper.tsx]: Just found a "bug" (not really a bug, moreso weird behavior). This component rerenders with page changes, which is probably not quite what we want. 

The way to fix this is to move this up in the component chain. Probably into `app.tsx` since there's not really much that's above the page wrapper (there's the router, but I wouldn't want this there).

If we do that, we're also gonna want to up our z-index of our loading overlay to 2 since on page transition we wouldn't want this popup to be seen above that overlay

**Review by MikeDrewitt — CHANGES_REQUESTED** [review]

File: src/components/pages/userDetailPage.scss (hunk 16 above)
Comment (Searay-330) [reply src/components/pages/userDetailPage.scss]: yeah ill move it we can make them look better later

File: src/components/shared/layouts/pageWrapper.tsx (hunk 19 above)
Comment (Searay-330) [reply src/components/shared/layouts/pageWrapper.tsx]: i moved it and it still re-renders on page change

File: src/components/shared/inputs/button.tsx [hunk 20]
```diff
@@ -0,0 +1,32 @@
+import React from 'react'
+
+import styles from './button.scss'
+
+type Props = {
+  label: string
```
Comment (MikeDrewitt) [inline src/components/shared/inputs/button.tsx]: This isn't a string, it's something else 

```suggestion
  children: React.ReactNode
```

This allows you to pass in it's child like 

```javascript
<Button>My button text!</Button>
```

instead

File: src/components/shared/inputs/button.tsx [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+  className?: string
+  onClick?: () => void
+  loading?: boolean
+}
+
+const Button = ({ className = '', label, loading = false, onClick }: Props) => {
+  if (loading)
```
Comment (MikeDrewitt) [inline src/components/shared/inputs/button.tsx]: You don't need two separate buttons here, you can just put the conditional within the jsx itself to clean this up a bit.

File: src/components/shared/inputs/button.tsx [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+    return (
+      <button
+        disabled
+        className={`${className} ${styles.defaultButton} ${styles.isLoading}`}
```
Comment (MikeDrewitt) [inline src/components/shared/inputs/button.tsx]: When passing a className from the parent, that classname should always be last here to allow any and all style overrides from that parent.

File: src/components/shared/inputs/faIconButton.tsx [hunk 23]
```diff
@@ -6,10 +6,14 @@ import styles from './faIconButton.scss'
 
 type Props = {
   onClick: () => void
+  className?: string
```
Comment (MikeDrewitt) [inline src/components/shared/inputs/faIconButton.tsx]: This doesn't need to be added here, it's already included in the props type from `FaIconProps` and `FaIconProps` is being extended. 

If you're looking for a way to style the button rather than the icon, then add a new prop called `buttonClass`, I think that'd be better cause then we can style the button and the icon if need be

File: src/components/shared/inputs/button.scss [hunk 24]
```diff
@@ -0,0 +1,57 @@
+@import 'variables';
+
+.defaultButton {
```
Comment (MikeDrewitt) [inline src/components/shared/inputs/button.scss:3]: You need to rerun `npm run format`

File: src/components/shared/inputs/textField.scss [hunk 25]
```diff
@@ -8,4 +10,14 @@
 
 .input {
   padding: 5px;
+  background-color: $background;
```
Comment (MikeDrewitt) [inline src/components/shared/inputs/textField.scss]: This isn't a good idea without changing the color css as well since you can't see the default black text in dark mode.

Honestly we're probably gonna want to update this in it's own PR cause this component needs work. 

We could do it in this PR, but I think it's already getting scope creepy. We might just want to revert this components css to what it was for now.

**Review by MikeDrewitt — CHANGES_REQUESTED** [review]

File: src/components/shared/inputs/button.scss [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+.loading {
+  font-size: 14px;
+  position: absolute;
+}
+
+.loading:after {
+  overflow: hidden;
+  display: block;
+  vertical-align: bottom;
+  -webkit-animation: ellipsis steps(4,end) 1000ms infinite;
```
Comment (MikeDrewitt) [inline src/components/shared/inputs/button.scss:42]: This only works in chrome unfortunately, so not a great solution

File: src/components/shared/inputs/textField.scss (hunk 25 above)
Comment (Searay-330) [reply src/components/shared/inputs/textField.scss]: i agree but the default text color in dark mode is white, with this reverted my boxes look terrible in dark mode so maybe do a seprate pr merge that then this.

File: src/components/shared/forms/editUserForm.tsx [hunk 27]
```diff
@@ -0,0 +1,54 @@
+import React, { useState } from 'react'
```
Comment (MikeDrewitt) [inline src/components/shared/forms/editUserForm.tsx]: I don't think I'd put this in shared

I think forms should be a child of `components`

Shared is more for stuff that's meant to be building blocks, this is less of a building block and more of a finished castle.

File: src/components/shared/forms/editUserForm.tsx [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  return (
+    <div>
+      <h1>User Information</h1>
+      <TextField
+        onChange={handleUpdatePreferredName}
+        label='Preferred Name'
+        id='preferredName'
+        placeholder={user.preferredName}
```
Comment (MikeDrewitt) [inline src/components/shared/forms/editUserForm.tsx]: I don't think this should be a placeholder, I think this should be it's defaultValue, and we'll have to add that as a prop to text field

**Review by MikeDrewitt — CHANGES_REQUESTED** [review]

File: src/components/shared/inputs/button.scss (hunk 26 above)
Comment (Searay-330) [reply src/components/shared/inputs/button.scss:42]: i use firefox to dev it so it def works in FF

Comment (Searay-330) [reply src/components/shared/inputs/button.scss:42]: i actually broke it on the last commit will fix in the next one

File: src/components/shared/layouts/pageWrapper.tsx (hunk 19 above)
Comment (MikeDrewitt) [reply src/components/shared/layouts/pageWrapper.tsx]: No it doesn't what you did is fine.

**Comment by Searay-330** [comment]
Cool has to be the first time I got told something I make works better than I thought

**Review by MikeDrewitt — APPROVED** [review]
